### ISSUE TITLE
feature request: forbid unsafe downcasting

### ISSUE THREAD

**Issue by jmacmahon** [typescript-eslint/typescript-eslint#3226]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Example**

This code contains an unsafe downcast, which results in a TypeError at runtime:
```ts
type ResultSuccess<T> = { success: true, data: T }
type ResultFailure<E> = { success: false, error: E }
type Result<T, E> = ResultSuccess<T> | ResultFailure<E>

const doSomethingErrory = (): Result<string, Error> => {
  return { success: false, error: new Error('something') }
}

const result = doSomethingErrory()
(result as ResultSuccess<string>).data.length
// ^ Will throw in TypeError as res.data is undefined
```

It would be nice if there was a lint rule that disallowed this.

**Comment by bradzacher** [comment]
Type assertions in general should be avoided at all costs because they are very broadly unsafe.

They are allowed in the language to improve the devx, but you shouldn't ever really need one as runtime checks are infinitely safer.

Checking for assignability is VERY difficult for us to do. TypeScript does not expose any utilities to do it, which means we would essentially have to re-implement the typechecker to do it.

Which puts this rule into the "impossible to implement" bucket.

Consider using the [`consistent-type-assertions`](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/consistent-type-assertions.md) rule to ban assertions altogether.

**Comment by jmacmahon** [comment]
Oh, `consistent-type-assertions` seems to be exactly the kind of thing I was looking for, thanks.
